mitogen/master.py

```diff
@@ -177,17 +192,27 @@ def _looks_like_script(path):
 
 
 def _py_filename(path):
+    """
+    Returns a tuple of a Python path (if the file looks Pythonic) and whether or not
+    the Python path is special. Special file paths/modules might only exist in memory
+    """
     if not path:
-        return None
+        return None, False
 
     if path[-4:] in ('.pyc', '.pyo'):
         path = path.rstrip('co')
 
     if path.endswith('.py'):
-        return path
+        return path, False
 
     if os.path.exists(path) and _looks_like_script(path):
-        return path
+        return path, False
+
+    basepath = os.path.basename(path)
+    if basepath in SPECIAL_FILE_PATHS:
+        return path, True
+
+    return None, False
```

Comment:
Is there a non-error situation that would cause the function to reach this point? In other words, is there a danger that this line would mask a mistake elsewhere in the code? If so, would it make more sense to raise an exception, e.g. ValueError?

Reply:
🤔 I don't think so. The goal of the function is just to check if a file path is a python file path or not, so it's ok if it returns `False`. There's logic after:
```
if not path:
            return
```
in some places the function is used, which handle what happens if the file isn't Python :)

Reply:
My concern was that if we *expect* a line of code never to be reached (e.g. the last line of this function), then we should either a) leave it out, or b) write something that fails loudly (e.g. `raise SomeException('Unexpected fall though')`).

We shouldn't silently return a best guess/plausible value.

So, what is the expected condition/circumstance that the final `return None, False` is meant to handle? A code comment to clear that up would make this function easier to understand.

Reply:
That's a good concern. I double-checked out the places that `_py_filename` is called, and it's all in `mitogen/master.py`, in the various `find` methods of the various ways Mitogen attempts to determine the source code of a module here:
```
get_module_methods = [
        DefectivePython3xMainMethod(),
        PkgutilMethod(),
        SysModulesMethod(),
        ParentEnumerationMethod(),
    ]
```
`_py_filename` is used to attempt to determine the file path of the name of whatever module Mitogen is attempting to load. I think perhaps the function's name could be a bit more descriptive, ie something like `_get_python_file_path`, but the function can also be used as a utility as well, so that might be why the name wasn't changed? I don't really have a strong preference on the name change myself. 

In every case that `_py_filename` is called, handling of an empty `path` result reminds me of how `go` does it, where the result of `_py_filename()` is checked in an `if` statement, and if it's none then the `find()` method returns None and then the next `find()` method is called. I see 2 options here:

1. Add a code comment to explain that `return None, False` means that the filename passed to `_py_filename` does not appear to be python, and code later will handle when this function returns `None`.
2. Change the end of the function to something that fails loudly, then in all places where `_py_filename` is called, handle the failed exception and then make the `find()` functions return at that point if the loud exception is called.

Comparing the 2 options, checking a return type is less expensive than throwing an exception and then handling that exception, so for that reason I'm inclined to go with option 1 given that these `find` functions are called for every python lib that Mitogen tries to load. WDYT?

Reply:
Based on some [adhoc doctests](https://github.com/moreati/mitogen/commit/69ef1980e1dfa223bd63e1e5c2468ce7a996c54c) I'm satisfied with changes to this function.